### ISSUE TITLE
Intellisense suggesting code improvement not supported for the project's C# version

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#58133]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Intellisense-suggesting-code-improvement/1603447)._

---
[severity:It bothers me. A fix would be nice]
I have this code:

public class MyClass
    {
        public int ResponseCode { get; set; }
        public IList<AnotherClass> ResponseArray { get; set; }
        public string ResponseString { get; set; }
    }
    MyClass instance = new MyClass();
VS Intellisense shows:
>'new' expression can be simplified

and upon selecting that action, it changes `new MyClass()` to just `new()`, so that the line of code becomes:
`MyClass instance = new();`
but then build fails with **error code CS8181**:
>'new' cannot be used with tuple type. Use a tuple literal expression instead.

I think Intellisense has this bug where it's recommending a code change that's not supported for current **C# version 6.0**.

Microsoft Visual Studio Professional 2019 Version 16.11.7
.Net framework 4.6.1

[stackoverflow question] (https://stackoverflow.com/questions/70210593/visual-studio-intellisense-shows-new-expression-can-be-simplified-but-that-fa)

---
#### Feedback Bot on 12/5/2021, 06:44 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by Sergio0694** [comment]
I'm hitting this as well, I have a .NET Standard 2.0 project and Roslyn keeps suggesting to change a `.Slice(int)` call on a `Span<T>` instance to use a range, but ranges are not available in the project so applying the fix results in a build error.

**Comment by sharwell** [comment]
> I think Intellisense has this bug where it's recommending a code change that's not supported for current C# version 6.0.

This feature is only active in projects configured to use C# 9.0 or higher:

https://github.com/dotnet/roslyn/blob/e91fd748259b54a34b60cba635e72ce736295f4d/src/Analyzers/CSharp/Analyzers/UseImplicitObjectCreation/CSharpUseImplicitObjectCreationDiagnosticAnalyzer.cs#L43-L45

Note that this check was active from the time this feature was first introduced, so there was never a release where implicit object creation was suggested for a project targeting C# 6.0. Can you provide more details about the steps leading to the invalid suggestion?

**Comment by sharwell** [comment]
> I'm hitting this as well, I have a .NET Standard 2.0 project and Roslyn keeps suggesting to change a `.Slice(int)` call on a `Span<T>` instance to use a range, but ranges are not available in the project so applying the fix results in a build error.

This feature is only active in projects configured to use C# 8.0 or higher:

https://github.com/dotnet/roslyn/blob/e91fd748259b54a34b60cba635e72ce736295f4d/src/Analyzers/CSharp/Analyzers/UseIndexOrRangeOperator/CSharpUseRangeOperatorDiagnosticAnalyzer.cs#L97-L101

Can you provide more details about the steps leading to the invalid suggestion? It might help to file this as a separate bug since it affects a different analyzer/scenario/version check.

**Comment by Sergio0694** [comment]
@sharwell The issue is that if you have a .NET Standard 2.0 project using C# >= 8, the parser will support the range/index syntax, but the ref assembly will not have those types in the BCL, as they're only available in .NET Standard 2.1. I mean I guess that a configuration like this isn't technically supported and therefore the false positive is fine here? I understand that fixing this would require Roslyn to not only check the C# version but also the presence of the type symbol in the compilation, which would make things more complex there for a somewhat niche scenario. Feel free to just disregard if that's the case 🙂

**Comment by ryzngard** [comment]
Going to close this. I looked at making this fix and yea... it ends up being more work than we should do for unsupported language version + tfm pairs. There's other places this would need to be fixed too :/ For now we'll just keep this behavior since it only occurs in unsupported scenarios

**Comment by sharwell** [comment]
@Sergio0694 Starting with #36924, `CSharpUseRangeOperatorDiagnosticAnalyzer` and `CSharpUseIndexOperatorDiagnosticAnalyzer` are gated on the existence of `System.Range` and `System.Index`, respectively.

**Comment by sharwell** [comment]
The original issue above is a duplicate of #52727. It's impossible for the code fix to be offered in a project targeting C# 6, which means the file was included in more than one project and triggered from a project targeting C# 9 or newer. This is exactly the scenario described by #52727.
